File: ios/BodyCompass/Sources/BodyCompassCore/GoalProjection.swift
import Foundation

public struct GoalProjection: Codable, Equatable, Sendable {
    public let currentFatMassKg: Double
    public let currentLeanMassKg: Double
    public let targetWeightKg: Double
    public let fatToLoseKg: Double
    public let optimumWeeks: Int
    public let aggressiveWeeks: Int
    public let conservativeWeeks: Int
    public let weeklyLossTargetKg: Double
    public let dailyDeficitKcal: Int
    public let status: ProjectionStatus
    public let explanation: String
}

public enum ProjectionStatus: String, Codable, Equatable, Sendable {
    case alreadyAtGoal
    case onTrack
    case needsBetterAdherence
    case unsafePace
}

public enum GoalProjectionError: Error, Equatable {
    case invalidWeight
    case invalidBodyFat
    case targetNotLowerThanCurrent
}

public struct GoalProjectionCalculator: Sendable {
    public init() {}

    public func project(profile: BodyProfile) throws -> GoalProjection {
        guard profile.weightKg > 0 else { throw GoalProjectionError.invalidWeight }
        guard profile.bodyFatPercentage > 0, profile.bodyFatPercentage < 70 else {
            throw GoalProjectionError.invalidBodyFat
        }

        if profile.bodyFatPercentage <= profile.targetBodyFatPercentage {
            let fatMass = profile.weightKg * profile.bodyFatPercentage / 100
            let leanMass = profile.weightKg - fatMass
            return GoalProjection(
                currentFatMassKg: fatMass,
                currentLeanMassKg: leanMass,
                targetWeightKg: profile.weightKg,
                fatToLoseKg: 0,
                optimumWeeks: 0,
                aggressiveWeeks: 0,
                conservativeWeeks: 0,
                weeklyLossTargetKg: 0,
                dailyDeficitKcal: 0,
                status: .alreadyAtGoal,
                explanation: "You are already at or below the target body-fat percentage. Focus on maintenance, strength, and recovery."
            )
        }

        guard profile.targetBodyFatPercentage > 3, profile.targetBodyFatPercentage < profile.bodyFatPercentage else {
            throw GoalProjectionError.targetNotLowerThanCurrent
        }

        let currentFatMass = profile.weightKg * profile.bodyFatPercentage / 100
        let currentLeanMass = profile.weightKg - currentFatMass
        let targetWeight = currentLeanMass / (1 - profile.targetBodyFatPercentage / 100)
        let fatToLose = max(0, profile.weightKg - targetWeight)

        let baseWeeklyLoss = min(max(profile.weightKg * 0.0075, 0.35), 0.85)
        let adherence = min(max(profile.adherenceScore, 0.35), 1)
        let observedTrend = profile.weeklyWeightTrendKg.map { abs($0) }
        let weeklyTarget = max(0.25, min(observedTrend ?? baseWeeklyLoss, baseWeeklyLoss) * adherence)

        let optimumWeeks = Int(ceil(fatToLose / weeklyTarget))
        let aggressiveWeeks = Int(ceil(fatToLose / min(profile.weightKg * 0.01, 1.0)))
        let conservativeWeeks = Int(ceil(fatToLose / max(profile.weightKg * 0.005, 0.25)))
        let dailyDeficit = Int((weeklyTarget * 7700 / 7).rounded())

        let status: ProjectionStatus
        if dailyDeficit > 1100 {
            status = .unsafePace
        } else if adherence < 0.65 {
            status = .needsBetterAdherence
        } else {
            status = .onTrack
        }

        return GoalProjection(
            currentFatMassKg: currentFatMass,
            currentLeanMassKg: currentLeanMass,
            targetWeightKg: targetWeight,
            fatToLoseKg: fatToLose,
            optimumWeeks: optimumWeeks,
            aggressiveWeeks: aggressiveWeeks,
            conservativeWeeks: conservativeWeeks,
            weeklyLossTargetKg: weeklyTarget,
            dailyDeficitKcal: dailyDeficit,
            status: status,
            explanation: "Projection assumes lean mass is mostly preserved through protein, resistance training, sleep, and a moderate calorie deficit."
        )
    }
}

File: server/src/services/goalProjection.js
export function createProjection(profile) {
  const weightKg = Number(profile.weightKg);
  const bodyFatPercentage = Number(profile.bodyFatPercentage);
  const targetBodyFatPercentage = Number(profile.targetBodyFatPercentage ?? 12);
  const adherenceScore = clamp(Number(profile.adherenceScore ?? 0.75), 0.35, 1);

  if (!Number.isFinite(weightKg) || weightKg <= 0) {
    throw new Error("weightKg must be a positive number");
  }

  if (!Number.isFinite(bodyFatPercentage) || bodyFatPercentage <= 0 || bodyFatPercentage >= 70) {
    throw new Error("bodyFatPercentage must be between 0 and 70");
  }

  if (bodyFatPercentage <= targetBodyFatPercentage) {
    return {
      currentFatMassKg: round(weightKg * bodyFatPercentage / 100),
      currentLeanMassKg: round(weightKg * (1 - bodyFatPercentage / 100)),
      targetWeightKg: round(weightKg),
      fatToLoseKg: 0,
      optimumWeeks: 0,
      aggressiveWeeks: 0,
      conservativeWeeks: 0,
      weeklyLossTargetKg: 0,
      dailyDeficitKcal: 0,
      status: "alreadyAtGoal",
      explanation: "You are already at or below target. Focus on maintenance and performance."
    };
  }

  const currentFatMassKg = weightKg * bodyFatPercentage / 100;
  const currentLeanMassKg = weightKg - currentFatMassKg;
  const targetWeightKg = currentLeanMassKg / (1 - targetBodyFatPercentage / 100);
  const fatToLoseKg = Math.max(0, weightKg - targetWeightKg);
  const baseWeeklyLossKg = clamp(weightKg * 0.0075, 0.35, 0.85);
  const observedTrendKg = profile.weeklyWeightTrendKg == null ? null : Math.abs(Number(profile.weeklyWeightTrendKg));
  const weeklyLossTargetKg = Math.max(0.25, Math.min(observedTrendKg ?? baseWeeklyLossKg, baseWeeklyLossKg) * adherenceScore);
  const dailyDeficitKcal = Math.round(weeklyLossTargetKg * 7700 / 7);

  return {
    currentFatMassKg: round(currentFatMassKg),
    currentLeanMassKg: round(currentLeanMassKg),
    targetWeightKg: round(targetWeightKg),
    fatToLoseKg: round(fatToLoseKg),
    optimumWeeks: Math.ceil(fatToLoseKg / weeklyLossTargetKg),
    aggressiveWeeks: Math.ceil(fatToLoseKg / Math.min(weightKg * 0.01, 1)),
    conservativeWeeks: Math.ceil(fatToLoseKg / Math.max(weightKg * 0.005, 0.25)),
    weeklyLossTargetKg: round(weeklyLossTargetKg),
    dailyDeficitKcal,
    status: adherenceScore < 0.65 ? "needsBetterAdherence" : "onTrack",
    explanation: "Projection assumes lean mass is mostly preserved through protein, strength training, sleep, and a moderate deficit."
  };
}

function clamp(value, min, max) {
  return Math.min(Math.max(value, min), max);
}

function round(value) {
  return Math.round(value * 10) / 10;
}
